Feat: SSO group sync

p-fruck opened the pull request:
Implementation of the feature discussed in https://github.com/inventree/InvenTree/discussions/7200

I was unable to test the new platform UI, I hope the modifications to `SystemSettings.tsx` are sufficient. Currently missing translations for the new settings. For some reason `python3 manage.py makemessages -a` doesn't seem to work inside the dev container?

p-fruck commented:
@wolflu05 thanks, I updated the code accordingly. Are there also plans towards integrating the frontend into the docker-compose setup?

SchrodingersGat commented on src/backend/InvenTree/common/models.py:
```diff
@@ -1930,6 +1930,32 @@ def save(self, *args, **kwargs):
+        },
+        'SSO_GROUP_KEY': {
+            'name': 'SSO group key',
+            'description': 'The name of the groups claim attribute provided by the IdP',
```
Please make these strings translateable, by using the `_('The name of the groups...')` approach

SchrodingersGat replied:
The "name" and "description" field for each should be wrapped like that.

p-fruck replied:
resolved in d6b6e241584c72905865e2176ea7212b9fbe89d1

matmair commented:
The dev container comes with yarn support, the docker-compose file is not meant to be a full dev environment as that enlargens the attack surface a lot. We are shipping the compile frontend with the docker container.

p-fruck commented:
> We are shipping the compile frontend with the docker container

@matmair are you referring to the dev or prod container? I cannot see the compiled frontend code being added to the dev container.
I guess I will play around with manual `yarn dev` as devcontainers do not work for me. Maybe if there is some possibility to launch them without using VSCode :smile:

matmair commented:
@p-fruck prod. The compiled frontend is in `/src/backend/InvenTree/web/static/web`. We are very much vscode oriented, most members of the core team use it so there is not much consideration given for other editors.

p-fruck commented:
Hey all, I rebased to master (nice shortcut for global settings :wink:) and fixed the links to the django allauth docs in [7dca6a6](https://github.com/inventree/InvenTree/pull/7293/commits/7dca6a6b206c04a40d72f907cc661a39bfca823e)
Is there anything left holding back this PR?

SchrodingersGat commented:
It would nice to be able to add unit tests for this but I'm not sure how we would achieve that. This looks good from my end. Any thoughts @matmair ?

matmair commented on src/backend/InvenTree/InvenTree/sso.py:
```diff
@@ -75,3 +82,53 @@ def registration_enabled() -> bool:
+    if get_global_setting('SSO_REMOVE_GROUPS'):
+        for group in user.groups.all():
+            if not group.name in group_names:
+                logger.warning(f'Removing group {group.name} from {user}')
```
Comment (blocking): In my opinion it would make more sense to log this as an info - this is an intended function and does not require remidiation by an adding. Thoughts @SchrodingersGat?

p-fruck replied:
Done :+1:

p-fruck commented on src/backend/InvenTree/InvenTree/sso.py:
```diff
@@ -75,3 +82,53 @@ def registration_enabled() -> bool:
+    # map SSO groups to InvenTree groups
+    group_names = []
+    for sso_group in sociallogin.account.extra_data[group_key]:
+        group_names.append(group_map[sso_group])
```
Thats correct, I've changed the code to check if the keys exist in `extra_data` and  `group_map`

p-fruck commented on src/backend/InvenTree/InvenTree/sso.py:
```diff
@@ -75,3 +82,53 @@ def registration_enabled() -> bool:
+                logger.info(f'Adding group {group_name} to user {user}')
+            except Group.DoesNotExist:
+                logger.error(
+                    f'Failed to add group {group_name} to user {user}: Group does not exist'
```
Honestly I don't see any difference here. An empty group that does not have permissions does not make sense in my opinion and if the group is created by an admin and assigned the correct permission, the user will be added to the group during the next signin anyway

matmair replied:
The user would not have to log off once the admin adds the missing permissions from the group but we can address this in a later PR at any point.

p-fruck replied:
I agree this might be a benefit. I've added the functionality and a unit test for it

p-fruck commented:
Thanks for your suggestions! I've added a bunch of unit tests and made some changes to the docs

matmair reviewed (approved):
LGTM; thank you for the tests. The backend CI error has occurred before, I do not think that is this PR.

SchrodingersGat commented:
@p-fruck thanks muchly for the contribution :)
